python/ray/autoscaler/_private/kubernetes_operator/node_provider.py

```diff
@@ -0,0 +1,297 @@
+
+logger = logging.getLogger(__name__)
+
+
+def to_label_selector(tags):
+    label_selector = ""
+    for k, v in tags.items():
+        if label_selector != "":
+            label_selector += ","
+        label_selector += "{}={}".format(k, v)
+    return label_selector
+
+
+class KubernetesOperatorNodeProvider(NodeProvider):
+    def __init__(self, provider_config, cluster_name):
+        NodeProvider.__init__(self, provider_config, cluster_name)
+        self.cluster_name = cluster_name
+        self.namespace = provider_config["namespace"]
+
+    def non_terminated_nodes(self, tag_filters):
+        # Match pods that are in the 'Pending' or 'Running' phase.
+        # Unfortunately there is no OR operator in field selectors, so we
+        # have to match on NOT any of the other phases.
+        field_selector = ",".join([
+            "status.phase!=Failed",
+            "status.phase!=Unknown",
+            "status.phase!=Succeeded",
+            "status.phase!=Terminating",
+        ])
+
+        tag_filters[TAG_RAY_CLUSTER_NAME] = self.cluster_name
+        label_selector = to_label_selector(tag_filters)
+        pod_list = core_api().list_namespaced_pod(
+            self.namespace,
+            field_selector=field_selector,
+            label_selector=label_selector)
+
+        return [pod.metadata.name for pod in pod_list.items]
+
+    def is_running(self, node_id):
+        pod = core_api().read_namespaced_pod(node_id, self.namespace)
+        return pod.status.phase == "Running"
+
+    def is_terminated(self, node_id):
+        pod = core_api().read_namespaced_pod(node_id, self.namespace)
+        return pod.status.phase not in ["Running", "Pending"]
+
+    def node_tags(self, node_id):
+        pod = core_api().read_namespaced_pod(node_id, self.namespace)
+        return pod.metadata.labels
+
+    def external_ip(self, node_id):
+        raise NotImplementedError("Must use internal IPs with Kubernetes.")
+
+    def internal_ip(self, node_id):
+        pod = core_api().read_namespaced_pod(node_id, self.namespace)
+        return pod.status.pod_ip
+
+    def get_node_id(self, ip_address, use_internal_ip=True) -> str:
+        if not use_internal_ip:
+            raise ValueError("Must use internal IPs with Kubernetes.")
+        return super().get_node_id(ip_address, use_internal_ip=use_internal_ip)
+
+    def set_node_tags(self, node_id, tags):
+        pod = core_api().read_namespaced_pod(node_id, self.namespace)
+        pod.metadata.labels.update(tags)
+        core_api().patch_namespaced_pod(node_id, self.namespace, pod)
+
+    def create_node(self, node_config, tags, count):
+        # create a raycluster
+        conf = node_config.copy()
+        pod_spec = conf.get("pod", conf)
+        service_spec = conf.get("service")
+        ingress_spec = conf.get("ingress")
+        node_uuid = str(uuid4())
+        tags[TAG_RAY_CLUSTER_NAME] = self.cluster_name
+        tags["ray-node-uuid"] = node_uuid
```

Comment:
Why not use a constant to place "ray-node-uuid"

Reply:
This rule follows the way in the other node provider implementation.